### PR TITLE
fix(NODE-5818): Add feature flagging to server selection logging

### PR DESCRIPTION
**Pull request by aditi-khare-mongoDB**
### Description
Add feature flagging to server selection logging, so that event instances are only created when `willLog` returns `true` for the logging configuration.

#### **EVG Perf Analysis Results**

Upon comparing the perf results of NODE-5818's changes to the commit before server selection logging was merged ([b93d405](https://evergreen.mongodb.com/task/mongo_node_driver_next_performance_tests_run_spec_benchmark_tests_node_18_server_6.0_b93d405275c3a8ce6b1eba0af335ffb2a309e34e_24_01_03_17_07_24), there are no significant regressions through this commit. 
Both commits were rerun at the same time to ensure no ci changes influence the results.

Here is the [Spreadsheet Link](https://docs.google.com/spreadsheets/d/1L_tlp-n5-tUFVuSAQrmIdC0aqMNvFsPLKqV8my-ACE8/edit#gid=0).

#### What is changing?
Same as above.

##### Is there new documentation needed for these changes?
No.

#### What is the motivation for this change?
Before: (Due to server selection logging merge [NODE-4687](https://jira.mongodb.org/browse/NODE-4687))
findOne - 6% throughput reduction
runCommand - 9% throughput reduction
smallDocInsertOne - 6.5% throughput reduction

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

**Comment by durran** [comment]
Am I correct in analysing this that, for example, `findOne` on these changes is slower than the mainline?

Screenshot 2024-01-24 at 20 02 30

**Comment by aditi-khare-mongoDB** [comment]
Bookkeeping: 
Screenshot 2024-01-24 at 3 13 03 PM

With newest merge to main, this is a perf improvement.

**Comment by aditi-khare-mongoDB** [comment]
Perf Analysis Results:

Evergreen Trends:
Upon comparing the perf results of NODE-5818's changes to the commit before server selection logging was merged ([b93d405](https://evergreen.mongodb.com/task/mongo_node_driver_next_performance_tests_run_spec_benchmark_tests_node_18_server_6.0_b93d405275c3a8ce6b1eba0af335ffb2a309e34e_24_01_03_17_07_24). There are no regressions as high as 6% through this commit, but one `FindAndEmptyCursor` regressed by 3% . All other tests are perf improvements or regressions by les than 1%

[Spreadsheet Link](https://docs.google.com/spreadsheets/d/1L_tlp-n5-tUFVuSAQrmIdC0aqMNvFsPLKqV8my-ACE8/edit#gid=0)

File: src/sdam/topology.ts
Comment (alenakhineika) [inline src/sdam/topology.ts:573]: (nit): Would it be beneficial to create a function similar to `shouldEmitAndLogCommand` and move the `willLog` check from here? True that there is only one condition that we check, but maybe with the function name it will look shorter and easier to read code. Plus, if we need to change it later, we will need to modify only one function rather than multiple `if`'s in the code. Looks good as is, just caught my eye since we are doing it quite a number of times in this pr.

Comment (aditi-khare-mongoDB) [reply src/sdam/topology.ts:573]: I was thinking something like that, but there are two differences between server selection logging and command logging to consider here:

1. Unlike CLAM, Server selection logging has different severity levels for its logs (debug and info), so regardless we will have to pass something into the function
2. in `shouldEmitAndLogCommand`, we check multiple booleans, here we are just calling one function.

**Review by alenakhineika — APPROVED** [review]

**Review by W-A-James — APPROVED** [review]

**Comment by alenakhineika** [comment]
The PR is approved by the team, but blocked by [DEVPROD-4523](https://jira.mongodb.org/browse/DEVPROD-4523) and can be merged only when the waterfall is fixed.

Also flaky tests: NODE-5897, NODE-5498, NODE-5803

**Review by durran — CHANGES_REQUESTED** [review]
I think we just need to rebase now, remove the `willLog` checks, and then make all calls to the mongologger user `topology.client.mongoLogger?.debug()` now.

**Comment by aditi-khare-mongoDB** [comment]
@durran what is your reasoning for removing the `willLog` / feature flagging checks?

**Comment by durran** [comment]
> @durran what is your reasoning for removing the `willLog` / feature flagging checks?

@aditi-khare-mongoDB The logger instantiation PR refactored the `log` method in the client logger to check `willLog` first and return if it won't log. `debug` is bound to `log` so it's going to check that first as well - I don't see the need for the redundant check to `willLog`. That's why I thought this PR would just go away.

**Comment by aditi-khare-mongoDB** [comment]
@durran 
I see, while it might seem that way the log method always checked `willLog` or some similar method (`compareSeverities`) before logging. See [this PR](https://github.com/mongodb/node-mongodb-native/pull/3971/files#diff-e3cf2d38cd60718791de47479d524b551c336cf17919e9f285249d62adb374d5) (a random logging PR from before the logger instantiation PR), for reference. 

The focus of this PR is so that a `ServerSelectionEvent` instance is only created if `willLog` returns true, and the log call only occurs after the `ServerSelectionEvent` instance is created. This is why, I am of the opinion that we have to add a willLog check before we call `MongoLogger?.debug `or `MongoLogger?.info`.

**Comment by durran** [comment]
> @durran I see. The `log` method always checked `willLog` or some similar method (`compareSeverities`) before logging the message. See [this PR](https://github.com/mongodb/node-mongodb-native/pull/3971/files#diff-e3cf2d38cd60718791de47479d524b551c336cf17919e9f285249d62adb374d5) (a random logging PR from before the logger instantiation PR), for reference.
> 
> The focus of this PR is so that a `ServerSelectionEvent` instance is only created if `willLog` returns true, and the `log` call only occurs after the `ServerSelectionEvent` instance is created. This is why, I am of the opinion that we should add a `willLog` check before we call `MongoLogger?.debug `or `MongoLogger?.info`.

Ah ok, thanks for clarifying that. Then I think we can proceed here.

**Review by alenakhineika — APPROVED** [review]

**Pull request merged**
